### PR TITLE
Initial QEMU Support

### PR DESCRIPTION
**Pull request by colpane**
A **qemu-image** is added for booting IOT2050 on qemu system. You can test the not hardware dependent functionality using **qemu-image**.

All your comments/suggestions/criticisms are welcomed.
Thanks..

### CODE REVIEW

**Comment by colpane** [comment]
@jan-kiszka @BaochengSu Please review and comment. :)

File: Kconfig [hunk 1]
```diff
@@ -116,7 +116,11 @@ config SECURE_BOOT
 	  This uses the public custMpk.key and certificate by default. Do not
 	  use this key in production, it is for demonstration purposes only.
 
-	  Note: The persistent /var and /home partitions are not yet protected!
+	  The writable persistent /var and /home partitions are encrypted as
+	  LUKS2 partitions.
+
+	  The reason of coupling the data encryption with the secure boot is to
+	  have a proper integrity chain rooted in hardware (OTP, RPMB, etc.)
```
Comment (BaochengSu) [inline Kconfig]: This could be rebased into the first commit.

File: Kconfig [hunk 2]
```diff
@@ -30,6 +30,11 @@ config IMAGE_SWUPDATE
 	  Based on the example image, this adds SWUpdate and changes the
 	  partition layout to an A/B rootfs.
 
+config IMAGE_QEMU
+	bool "QEMU Image"
+	help
+	  QEMU image to boot image without real hardware.
```
Comment (jan-kiszka) [inline Kconfig:36]: Should mention the differences to the real device, e.g. the missing Arduino I/Os.

File: Kconfig [hunk 3]
```diff
@@ -110,11 +116,7 @@ config SECURE_BOOT
 	  This uses the public custMpk.key and certificate by default. Do not
 	  use this key in production, it is for demonstration purposes only.
 
-	  The writable persistent /var and /home partitions are encrypted as
-	  LUKS2 partitions.
-
-	  The reason of coupling the data encryption with the secure boot is to
-	  have a proper integrity chain rooted in hardware (OTP, RPMB, etc.)
+	  Note: The persistent /var and /home partitions are not yet protected!
```
Comment (jan-kiszka) [inline Kconfig]: Unrelated changes.

File: conf/distro/iot2050-debian-qemu.conf [hunk 4]
```diff
@@ -0,0 +1,15 @@
+#
+# Copyright (c) Siemens AG, 2023
+#
+# Authors:
+#  Enes �olpan <enes.colpan@siemens.com>
```
Comment (jan-kiszka) [inline conf/distro/iot2050-debian-qemu.conf:5]: Is this UTF8 in your name and just a limitation of github in the visualization?

File: conf/distro/iot2050-debian-qemu.conf [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+#
+# This file is subject to the terms and conditions of the MIT License.  See
+# COPYING.MIT file in the top-level directory.
+#
+
+require conf/distro/iot2050-debian.conf
+
+# used for SWUpdate boot
+ABROOTFS_IMAGE_RECIPE = "iot2050-image-qemu"
+VERITY_IMAGE_RECIPE = "iot2050-image-qemu"
```
Comment (jan-kiszka) [inline conf/distro/iot2050-debian-qemu.conf:15]: Do we really need an own image, or can we also get away with just an own machine?

File: recipes-core/images/iot2050-image-qemu.bb [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+# EFI Boot Guard is used instead
+IMAGE_INSTALL:remove = "u-boot-script"
+
+IMAGE_INSTALL += "efibootguard"
+IMAGE_INSTALL += "swupdate"
+IMAGE_INSTALL += "swupdate-handler-roundrobin"
+IMAGE_INSTALL += "swupdate-complete-update-helper"
+IMAGE_INSTALL += "iot2050-watchdog"
+
+DTB_FILES = ""
```
Comment (jan-kiszka) [inline recipes-core/images/iot2050-image-qemu.bb:42]: This is really too much duplication and will be hard to maintain, constantly broken. Focus on the real differences.

File: recipes-kernel/linux/files/iot2050_defconfig_base [hunk 7]
```diff
@@ -691,3 +691,8 @@ CONFIG_SAMPLES=y
 CONFIG_SAMPLE_RPMSG_CLIENT=m
 CONFIG_MEMTEST=y
 CONFIG_TI_DAVINCI_MDIO=y
+CONFIG_SERIAL_AMBA_PL011=y
+CONFIG_SERIAL_AMBA_PL011_CONSOLE=y
+CONFIG_SCSI_VIRTIO=y
+CONFIG_DRM_VIRTIO_GPU=y
+CONFIG_VIRTIO_INPUT=y
```
Comment (jan-kiszka) [inline recipes-kernel/linux/files/iot2050_defconfig_base:698]: Missing newline at end of file.

And this defconfig extension should probably go into the extra set as it is not need for the IOT2050 base.

Comment (jan-kiszka) [reply recipes-kernel/linux/files/iot2050_defconfig_base:698]: ..and that as an own commit.

File: wic/iot2050-qemu.wks.in [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+include ebg-sysparts.inc
+
+part --source rawcopy --sourceparams "file=${IMAGE_FULLNAME}.squashfs" --align 1024 --fixed-size 4G --uuid "${ABROOTFS_PART_UUID_A}"
+part --source empty --align 1024 --fixed-size 4G --uuid "${ABROOTFS_PART_UUID_B}"
+
+# persistent partitions
+part /var --source rootfs --rootfs-dir=${IMAGE_ROOTFS}/var --label var --fstype=ext4 --mkfs-extraopts "-T default" --ondisk sda --align 1024  --size 512M
+part /home --source rootfs --rootfs-dir=${IMAGE_ROOTFS}/home --label home --fstype=ext4 --mkfs-extraopts "-T default" --ondisk sda --align 1024  --size 512M
+
+bootloader --ptable gpt --append "rootwait ${EXTRA_KERNEL_PARAMS}"
```
Comment (jan-kiszka) [inline wic/iot2050-qemu.wks.in:20]: Again, too much duplication. Why do we need an own image layout?

File: Kconfig (hunk 1 above)
Comment (jan-kiszka) [reply Kconfig]: First commit was wrong already, and fixing that will obsolete this one.

File: conf/distro/iot2050-debian-qemu.conf (hunk 4 above)
Comment (BaochengSu) [reply conf/distro/iot2050-debian-qemu.conf:5]: I guess this is a typo -- see recipes-core/images/iot2050-image-qemu.bb

File: wic/iot2050-qemu.wks.in (hunk 8 above)
Comment (colpane) [reply wic/iot2050-qemu.wks.in]: Removed file and used swu.wks file with a variable.

File: recipes-kernel/linux/files/iot2050_defconfig_base (hunk 7 above)
Comment (colpane) [reply recipes-kernel/linux/files/iot2050_defconfig_base]: Done.

File: conf/distro/iot2050-debian-qemu.conf (hunk 5 above)
Comment (colpane) [reply conf/distro/iot2050-debian-qemu.conf]: Switched to machine instead of image.

File: conf/distro/iot2050-debian-qemu.conf (hunk 4 above)
Comment (colpane) [reply conf/distro/iot2050-debian-qemu.conf]: File is deleted, no problem anymore :)

File: Kconfig (hunk 3 above)
Comment (colpane) [reply Kconfig]: There was a rebase issue, now it is aligned with master branch.

File: Kconfig (hunk 1 above)
Comment (colpane) [reply Kconfig]: There was a rebase issue, now it is aligned with master branch.

File: Kconfig (hunk 2 above)
Comment (colpane) [reply Kconfig:36]:
```suggestion
	  QEMU image to boot image without real hardware. This image has
          missing or not working features especially dependent on hardware.
          Please consider limited functionality before using this (e.g. the
          missing Arduino I/Os).
```

Is this information enough/clear? What could I add more?

File: conf/machine/iot2050-qemu.conf [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+#
+
+require conf/machine/iot2050.conf
+
+DTB_FILES = ""
+
+CONSOLE_KERNEL_PARAMS = ""
+
+PREFERRED_PROVIDER_u-boot-${MACHINE} = "u-boot-qemu-arm64"
+PREFERRED_PROVIDER_u-boot-${MACHINE}-config = "u-boot-qemu-arm64"
```
Comment (colpane) [inline conf/machine/iot2050-qemu.conf:18]: Somehow this generates u-boot-pg1 instead of qemu. Will fix it.

File: Kconfig (hunk 2 above)
Comment (jan-kiszka) [reply Kconfig:36]: What else is different? We only have one Ethernet port configured via the command line below. Only one UART, the debug one.

I would write it like this: "The device emulated by QEMU is not identical to a physical IOT2050. This leads to missing features such as ... Please be aware of potential problems when using the image for testing purposes."

File: README.md [hunk 10]
```diff
@@ -58,6 +58,28 @@ The SDK is also available as docker image. To import it into a docker host, run
...
+
+```shell
+./kas-container menu
+```
+
+Then below command can be used to boot qemu image on a platform that
+qemu-system-aarch64 is installed.
+
+```shell
+qemu-system-aarch64 -bios ./firmware.bin -m 2G -netdev user,id=net,hostfwd=tcp:127.0.0.1:22222-:22
```
Comment (jan-kiszka) [inline README.md:74]: I suspect there is ./firmware.bin after building the QEMU image, is there? Better reference the one in the deploy folder.

File: conf/machine/iot2050-qemu.conf (hunk 9 above)
Comment (jan-kiszka) [reply conf/machine/iot2050-qemu.conf:18]: Hmm, where should that qemu u-boot come from? A Debian package?

Comment (colpane) [reply conf/machine/iot2050-qemu.conf:18]: It comes from isar-cip-core repository.
https://gitlab.com/cip-project/cip-core/isar-cip-core/-/tree/master/recipes-bsp/u-boot

File: README.md (hunk 10 above)
Comment (colpane) [reply README.md:74]: It will be fixed with upper u-boot bug. Then I will reference the exact path.

File: conf/machine/iot2050-qemu.conf (hunk 9 above)
Comment (jan-kiszka) [reply conf/machine/iot2050-qemu.conf:18]: Ah, of course!

File: Kconfig [hunk 11]
```diff
@@ -58,10 +63,11 @@ config KAS_INCLUDE_MAIN
 	default "kas-iot2050-swupdate.yml" if IMAGE_SWUPDATE
 	default "kas-iot2050-boot-pg1.yml" if IMAGE_BOOT_PG1
 	default "kas-iot2050-boot-pg2.yml" if IMAGE_BOOT_PG2
+	default "kas-iot2050-qemu.yml" if IMAGE_QEMU
```
[the remark this answers is not in the document]
Comment (colpane) [reply Kconfig:66]: Swupdate support is already included in qemu. The reference for qemu was the image with swupdate support. Should we have two different qemu image for swupdate image and w/o swupdate example image?

[the remark this answers is not in the document]
Comment (jan-kiszka) [reply Kconfig:66]: I see now. I would suggest to permit both image variants. Could be achieved by making the qemu yaml an option file that just switches the machine.

[the remark this answers is not in the document]
Comment (colpane) [reply Kconfig:66]: Yes that would be definitely better. I will check and change the related parts.

File: Kconfig (hunk 2 above)
Comment (colpane) [reply Kconfig]: Done.

File: conf/machine/iot2050-qemu.conf (hunk 9 above)
Comment (colpane) [reply conf/machine/iot2050-qemu.conf]: Fixed.

File: Kconfig [hunk 12]
```diff
@@ -95,7 +95,7 @@ config KAS_INCLUDE_LXDE
 	string
 	default "kas/opt/lxde.yml"
 	depends on LXDE
-
+
```
Comment (jan-kiszka) [inline Kconfig:98]: Stray whitespace chance.

File: recipes-core/images/iot2050-image-base.bb [hunk 13]
```diff
@@ -12,7 +12,7 @@ inherit image
 
 DESCRIPTION = "IOT2050 Debian Base Image"
 
-IMAGE_INSTALL += "u-boot-iot2050-config"
+IMAGE_INSTALL += "${@('u-boot-' + d.getVar('MACHINE')) if d.getVar('QEMU_IMAGE') == '1' else 'u-boot-${MACHINE}-config' }"
```
Comment (jan-kiszka) [inline recipes-core/images/iot2050-image-base.bb:15]: u-boot-*-config is needed for interacting with the U-Boot env. I don't think we have a persistent env with U-Boot under QEMU. In any case, we don't need u-boot-<machine> in that case.

File: recipes-core/images/iot2050-image-swu-example.bb [hunk 14]
```diff
@@ -40,6 +40,6 @@ IMAGE_INSTALL += "efibootguard libebgenv0"
 IMAGE_INSTALL += "customizations-swupdate"
 IMAGE_INSTALL += "swupdate-handler-roundrobin"
 IMAGE_INSTALL += "swupdate-complete-update-helper"
-IMAGE_INSTALL += "iot2050-watchdog"
+IMAGE_INSTALL += "${@ '' if d.getVar('QEMU_IMAGE') == '1' else 'iot2050-watchdog' }"
```
Comment (jan-kiszka) [inline recipes-core/images/iot2050-image-swu-example.bb:43]: Will things break if that is left in? But, yeah, we are missing some watchdog, see also https://gitlab.com/cip-project/cip-core/isar-cip-core/-/issues/66

Comment (colpane) [reply recipes-core/images/iot2050-image-swu-example.bb:43]: I think it will not be a problem. However, I thought that removing the package for qemu would make no watchdog support feature more visible. What do you suggest at the end?

Comment (jan-kiszka) [reply recipes-core/images/iot2050-image-swu-example.bb:43]: Change as little as needed to get a system starting that shows no errors.

File: recipes-core/images/iot2050-image-base.bb (hunk 13 above)
Comment (colpane) [reply recipes-core/images/iot2050-image-base.bb]: A new yml file is created for compiling u-boot qemu version. Removed from image.

File: recipes-core/images/iot2050-image-swu-example.bb (hunk 14 above)
Comment (colpane) [reply recipes-core/images/iot2050-image-swu-example.bb]: recipes-app/board-conf-tools/board-conf-tools_0.1.bb
This package could be removed as it gives error at the end of start process. However, it does not ruin booting process.

Comment (jan-kiszka) [reply recipes-core/images/iot2050-image-swu-example.bb]: Right, but that would be a good example for removing something that cannot work (given the existing device model in QEMU). Here, I would vote for removing, also to avoid confusion.

File: Kconfig [hunk 15]
```diff
@@ -50,6 +50,14 @@ config IMAGE_BOOT_PG2
...
 	  allows to recover. Otherwise you risk to BRICK THE IOT2050!
 
+config IMAGE_BOOT_QEMU
+	bool "Firmware image for QEMU"
+	help
+	  Build the firmware image to use with QEMU.
+
+	  WARNING: Do not flash this image onto your device. Otherwise you
+	  risk to BRICK THE IOT2050!
+
```
Comment (jan-kiszka) [inline Kconfig:60]: Nope, please not. Rather model that similar to isar-cip-core where U-Boot forms a helper for the QEMU start script. Maybe even consider a start script over lengthy README.

Comment (colpane) [reply Kconfig:60]: Sure, I will write a start-qemu script. This is for generating firmware.bin file for qemu, so we will keep this image type right?

Comment (jan-kiszka) [reply Kconfig:60]: https://gitlab.com/cip-project/cip-core/isar-cip-core/-/blob/master/kas/opt/ebg-swu.yml#L33 - this is how the u-boot firmware image is hooked up in isar-cip-core.

Comment (colpane) [reply Kconfig:60]: If we switch to this model, then we have to remove iot2050-watchdog. Otherwise, there is a conflict for compiling u-boot for qemu.

File: recipes-core/images/iot2050-image-swu-example.bb (hunk 14 above)
Comment (colpane) [reply recipes-core/images/iot2050-image-swu-example.bb:43]: Done.

File: Kconfig (hunk 15 above)
Comment (colpane) [reply Kconfig]: Done, switched that model and removed iot2050-watchdog.

File: start-qemu-iot2050.sh [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+	KERNEL_FILE=$(ls ${IMAGE_PREFIX}-vmlinu* | tail -1)
+	INITRD_FILE=$(ls ${IMAGE_PREFIX}-initrd.img* | tail -1)
+
+	${QEMU_PATH}${QEMU} \
+		-drive file=${IMAGE_FILE},discard=unmap,if=none,id=disk,format=raw \
+		-kernel ${KERNEL_FILE} -append "${KERNEL_CMDLINE}" \
+		-initrd ${INITRD_FILE} \
+		${QEMU_COMMON_OPTIONS} "$@"
+fi
+
```
Comment (jan-kiszka) [inline start-qemu-iot2050.sh:82]: Extra new-line at EOF.

**Comment by jan-kiszka** [comment]
Looks really good, just two style details! Will also give it a try soon, but that should not delay things.

**Comment by colpane** [comment]
Thank you Jan, especially for your valuable comments/suggestions. Will check last two style things.

**Comment by jan-kiszka** [comment]
When addressing remarks on patches, fold the changes into the affected patches, don't stack them on top.

**Comment by colpane** [comment]
> When addressing remarks on patches, fold the changes into the affected patches, don't stack them on top.

Like having separate commits for each goal?

**Comment by jan-kiszka** [comment]
That is the foundation of a patch series in any case. And if you modify something in v2 or later, only add patches if you address new separate topics.

**Comment by jan-kiszka** [comment]
Quick test feedback: Normal image works, but the SWUpdate / UEFI variant seems to get stuck past the kernel's EFI stub (no output on virtual UART). Does that work for you?

**Comment by colpane** [comment]
> Quick test feedback: Normal image works, but the SWUpdate / UEFI variant seems to get stuck past the kernel's EFI stub (no output on virtual UART). Does that work for you?

Yes, both of them were fine for me.

**Comment by jan-kiszka** [comment]
Yes, I'm using the script.

Hmm, now it works: Switched to the normal image, then back to the swupdate one, and now it's fine. Need to perform a clean build into the latter tomorrow to really rule out dependency issues.

**Comment by jan-kiszka** [comment]
All still good after a cleansstate rebuild. Must have been a partial rebuild fallout.

**Comment by jan-kiszka** [comment]
But now I'm testing secure boot + encryption. I'm not seeing TPM support in your start script, though... Maybe that feature should be grayed out for now?

**Comment by colpane** [comment]
You are right, I should gray out it if IMAGE_QEMU is selected. Also, I haven't test secure boot + encryption supported image.

File: Kconfig [hunk 17]
```diff
@@ -148,11 +148,29 @@ config RPMB_SETUP
...
+config KAS_INCLUDE_QEMU_IMAGE
+	string
+	default "kas-iot2050-qemu.yml"
+	depends on IMAGE_QEMU
 
 config KAS_INCLUDE_SECURE_BOOT
 	string
 	default "kas/opt/secure-boot.yml"
-	depends on SECURE_BOOT
+	depends on SECURE_BOOT && !IMAGE_QEMU
```
Comment (jan-kiszka) [inline Kconfig:173]: This is redundant as `SECURE_BOOT` has this precondition already.

**Review by jan-kiszka — CHANGES_REQUESTED** [review]
Looks good to me otherwise.

File: Kconfig (hunk 17 above)
Comment (colpane) [reply Kconfig]: Done.

**Comment by colpane** [comment]
Comments/problems are fixed and squashed unnecessary commits. Ready to merge 👍
@jan-kiszka @BaochengSu

File: wic/iot2050-swu.wks.in [hunk 18]
```diff
@@ -17,4 +17,4 @@ part --source empty --align 1024 --fixed-size 4G --uuid "${ABROOTFS_PART_UUID_B}
 part /var --source rootfs --rootfs-dir=${IMAGE_ROOTFS}/var --label var --fstype=ext4 --mkfs-extraopts "-T default" --ondisk sda --align 1024  --size 512M
 part /home --source rootfs --rootfs-dir=${IMAGE_ROOTFS}/home --label home --fstype=ext4 --mkfs-extraopts "-T default" --ondisk sda --align 1024  --size 512M
 
-bootloader --ptable gpt --append "console=ttyS3,115200n8 earlycon=ns16550a,mmio32,0x02810000 rootwait watchdog.handle_boot_enabled=0 ${EXTRA_KERNEL_PARAMS}"
+bootloader --ptable gpt --append "${CONSOLE_KERNEL_PARAMS} rootwait watchdog.handle_boot_enabled=0 ${EXTRA_KERNEL_PARAMS}"
```
Comment (jan-kiszka) [inline wic/iot2050-swu.wks.in:20]: Is that a "no new line at end of file"?

**Review by jan-kiszka — APPROVED** [review]
Just a style nit, looks good.

File: wic/iot2050-swu.wks.in (hunk 18 above)
Comment (BaochengSu) [reply wic/iot2050-swu.wks.in:20]: @colpane , If this is fixed then we can merge it :)

Comment (colpane) [reply wic/iot2050-swu.wks.in:20]: I can fix it, however all the wks.in files do not have a new line at the end of file and this wks.in file also did not have a new line at the end of file before I changed it :) Do I still need to add a new line?

Comment (jan-kiszka) [reply wic/iot2050-swu.wks.in:20]: The affected file currently does have a newline at its end.

Comment (colpane) [reply wic/iot2050-swu.wks.in:20]: Correct, it does not have, but other wks.in files (https://github.com/siemens/meta-iot2050/blob/master/wic/iot2050-swu-secure.wks.in) do not have it too. Why only this file should have the new line?

Comment (jan-kiszka) [reply wic/iot2050-swu.wks.in:20]: Just don't remove the newline removal from your patch and resubmit. Avoid style regressions.

Comment (colpane) [reply wic/iot2050-swu.wks.in]: Ah, sorry now I see the icon :) Thanks, fixed.

**Review by BaochengSu — APPROVED** [review]
